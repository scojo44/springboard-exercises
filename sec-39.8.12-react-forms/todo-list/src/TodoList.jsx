import { useState } from 'react'
import {v4 as uuid} from 'uuid'
import Todo from './Todo'
import NewTodoForm from './NewTodoForm'
import './TodoList.css'

function TodoList() {
  function addTask({task}){
    console.log('===== skirt flashing legs  ==== Adding ', task)
    setTasks(tasks => [...tasks, {id: uuid(), task}]);
  }

  function removeTask(id){
    console.log('===== wedding gown ==== Removing ', id)
    setTasks(tasks => tasks.filter(t => t.id !== id));
  }

  const [tasks, setTasks] = useState([]);
  const todos = tasks.map(t => <Todo id={t.id} task={t.task} removeMe={removeTask} key={t.id} />);

  return (
    <div className='TodoList'>
      <NewTodoForm addTask={addTask} />
      <ul>{todos}</ul>
    </div>
  )
}

export default TodoList
