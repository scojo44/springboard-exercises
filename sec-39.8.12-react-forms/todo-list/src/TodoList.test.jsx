import { render, fireEvent, getByDisplayValue } from '@testing-library/react';
import TodoList from './TodoList';

describe('TodoList Tests', () => {
  it('renders without crashing', () => {
    render(<TodoList />);
  });

  it('matches snapshot', () => {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('TodoList - Add/Remove Todo Tests', () => {
  function createTodo() {
    const {container, getByText, queryByText, getByLabelText, getByDisplayValue} = render(<TodoList />);
    const task = getByLabelText('New Task:');
    const button = getByText('Add Todo');
    
    // Confirm no todos displayed
    expect(container.querySelectorAll('.Todo').length).toBe(0);
    
    // Create a todo
    fireEvent.change(task, {target: {value: 'Learn TypeScript'}});
    fireEvent.click(button);
    
    return {container, getByText, queryByText, getByLabelText, getByDisplayValue};
  }
  
  it('creates a new todo and adds it to the list', () => {
    const {container} = createTodo();
    const todo = container.querySelector('.Todo');

    expect(container.querySelectorAll('.Todo').length).toBe(1);
    expect(container.querySelector('.TodoList ul')).toContainElement(todo);
    expect(todo).toHaveTextContent('Learn TypeScript');
  })

  it('should make it through the toggle process', () => {
    const {getByText, getByLabelText} = createTodo();
    const label = getByText('Learn TypeScript');
    const checkbox = getByLabelText('Learn TypeScript');

    // New task should not be marked as done
    expect(checkbox).not.toBeChecked();
    expect(label).not.toHaveClass('done');

    // Check the box to mark as completed
    fireEvent.click(checkbox);
    expect(checkbox).toBeChecked();
    expect(label).toHaveClass('done');

    // Check the box again to unmark as completed
    fireEvent.click(checkbox);
    expect(checkbox).not.toBeChecked();
    expect(label).not.toHaveClass('done');
  });

  it('should make it through the task edit process', () => {
    const {container, getByText, queryByText, getByDisplayValue} = createTodo();
    const editButton = getByText('Edit');

    // Bring up the edit UI
    fireEvent.click(editButton);
    expect(queryByText('Edit')).toBeNull();
    expect(queryByText('X')).toBeNull();
    const input = getByDisplayValue('Learn TypeScript');
    const saveButton = getByText('Save');

    // Save the updated task
    fireEvent.change(input, {target: {value: 'Learn Next.js'}});
    fireEvent.click(saveButton);

    // Verify back to view UI
    expect(queryByText('Save')).toBeNull();
    expect(queryByText('Edit')).not.toBeNull();
    expect(queryByText('X')).not.toBeNull();
    const todo = container.querySelector('.Todo');
    expect(todo).toHaveTextContent('Learn Next.js');
  });

  it('removes the new todo when the X button is clicked', () => {
    const {container, getByText} = createTodo();
    const removeButton = getByText('X');

    fireEvent.click(removeButton);

    expect(container.querySelectorAll('.Todo').length).toBe(0);
  });
});
