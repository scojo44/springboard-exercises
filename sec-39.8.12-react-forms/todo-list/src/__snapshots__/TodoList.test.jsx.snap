// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList Tests > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form
      class="NewTodoForm"
    >
      <label
        for="task"
      >
        New Task: 
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
